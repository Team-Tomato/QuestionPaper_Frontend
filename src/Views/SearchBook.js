import React, { Component } from 'react';
import '../Styles/style.css'
import Loader from 'react-loading';
import { Form, Row, Col, Container, Card, CardBody, Table } from 'reactstrap'
import ReactPaginate from 'react-paginate';
import '../Styles/pagination.css'
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faSortAlphaDown } from '@fortawesome/free-solid-svg-icons'

const StyledTextField = styled(TextField)`{
  label.Mui-focused {
    color: violet;
  }
  .MuiOutlinedInput-root {
    &.Mui-focused fieldset {
      border-color: violet;
    }
  }`;

const valreg=RegExp(/^\s+$/)

class Project extends Component {
  constructor(){
    super()
  this.state = {
    query: '',
    loading: false,
    person: [],
    offset:0,
    perPage:10,
    currentPage:0,
    noData:false,
    sort:false,
    choice:'',
    prev:''
  };
  this.handlePageClick = this.handlePageClick.bind(this)
  this.handleSubClick=this.handleSubClick.bind(this)
}
  onChange = e => {
    const { value } = e.target;
    let error= this.state.error;  
    if (value.length === 0 || valreg.test(value))
    {
      error = true;
    }
    else
    {
      error = false;
    }        
    this.setState({error,value },()=>console.log(this.state)) ;    
    this.setState({
      query: value,
      person: []
    });
  };

  handlePageClick=(e)=>{
    const selectedPage = e.selected;
    const offset = selectedPage * this.state.perPage;

    this.setState({
        currentPage: selectedPage,
        offset: offset
    });
  }
 
  async search(query) {
    const url = "https://teamtomato.herokuapp.com/api/v1/book/search?search_str=" + query //fetch the specific book
    const response = await fetch(url)
    const data = await response.json();
    this.setState({
      person: data,
      loading: false,
      sort:false,
      choice:'',
      prev:''
    })
    if(this.state.person.length==0)
    {
      this.setState({noData:true})
    }else{this.setState({noData:false})}
  }

  handleLogin = e => {
    const query = this.state.query
    this.setState({ loading: true })
    this.search(query)
  }

  handleSubClick=(e,value)=>{
    e.preventDefault();
  this.setState({sort:true,
    prev:this.state.choice,
  choice:value})
  }
  render() {
    const {error} = this.state;
    let BookContainer
    let table = []
    if (this.state.loading === false) {
      if (this.state.person !== [] && (this.state.person).length !== 0) {
        if(this.state.sort)
        {
          if(this.state.prev !== this.state.choice)
          {
          const v=this.state.choice
          const sorted=this.state.person.sort(function(a,b){
            if(a[v]<b[v])
            return -1;
            if(a[v]<b[v])
            return 1;
            return 0;
        }); 
        const slice=sorted.slice(this.state.offset, this.state.offset + this.state.perPage) 
        table = slice.map((data, index) => {
          return (
            <tr key={index}>
            <td>{data.title}</td>
            <td style={{textAlign:"center"}}>{data.author}</td>
            <td>{data.isbn}</td>
            <td style={{textAlign:"center"}}>{data.publisher}</td>
            <a href={data.url} target="blank" className="violet"><td>{data.url}</td></a>
          </tr>
          )
        })}
        else{
          const v=this.state.choice
          const sorted=this.state.person.sort(function(a,b){
            if(a[v]>b[v])
            return -1;
            if(a[v]>b[v])
            return 1;
            return 0;
        }); 
        const slice=sorted.slice(this.state.offset, this.state.offset + this.state.perPage) 
        table = slice.map((data, index) => {
          return (
            <tr key={index}>
            <td>{data.title}</td>
            <td style={{textAlign:"center"}}>{data.author}</td>
            <td>{data.isbn}</td>
            <td style={{textAlign:"center"}}>{data.publisher}</td>
            <a href={data.url} target="blank" className="violet"><td>{data.url}</td></a>
          </tr>
          )
        })
        }
        }
        else{
        const slice=this.state.person.slice(this.state.offset, this.state.offset + this.state.perPage)
        table = slice.map((data, index) => {
          return (
            <tr key={index}>
              <td>{data.title}</td>
              <td style={{textAlign:"center"}}>{data.author}</td>
              <td>{data.isbn}</td>
              <td style={{textAlign:"center"}}>{data.publisher}</td>
              <a href={data.url} target="blank" className="violet"><td>{data.url}</td></a>
            </tr>
          )
        })}
        BookContainer =
          <Container>
            <Table striped hover responsive>
              <thead>
                <tr>
                  <th><Button onClick={(event) => { this.handleSubClick(event, 'title')}} style={{backgroundColor: "violet",color:"white",width:100}}>Title <FontAwesomeIcon icon = {faSortAlphaDown}/></Button></th>
                  <th><Button onClick={(event) => { this.handleSubClick(event, 'author')}} style={{backgroundColor: "violet",color:"white",width:100}}>Author <FontAwesomeIcon icon = {faSortAlphaDown}/></Button> </th>
                  <th><Button onClick={(event) => { this.handleSubClick(event, 'isbn')}} style={{backgroundColor: "violet",color:"white",width:120}}>ISBN <FontAwesomeIcon icon = {faSortAlphaDown}/></Button></th>
                  <th><Button onClick={(event) => { this.handleSubClick(event, 'publisher')}} style={{backgroundColor: "violet",color:"white",width:120}}>Publisher <FontAwesomeIcon icon = {faSortAlphaDown}/></Button></th>
                  <th>URL</th>
                </tr>
              </thead>
              <tbody>
                {table}
              </tbody>
            </Table>
          </Container>
      }
      else {
        BookContainer = <h6></h6>
      }
    }
    else {
      BookContainer = <div style={{
        position: 'absolute', left: '50%', top: '77%',
        transform: 'translate(-50%, -50%)'
      }}>
        <Loader type={"bars"} color={"black"} />
      </div>
    }
    const count=Math.ceil(this.state.person.length / this.state.perPage);
    return (
      <div>
        <Card className="gradient">
          <CardBody className="welcome-title" style={{
            position: 'absolute', left: '50%', top: '50%',
            transform: 'translate(-50%, -50%)',
            color: 'white'
          }}>
            <h5>Books destination for Integrated students</h5>
          </CardBody>
        </Card>
        <br />
        <Container>
          <Card className="correctMargin">
            <CardBody className="removeIndent">
              <Form>
                <Row>
                  <Col sm={6} md={8} lg={10} className="addIndent">
                  <StyledTextField fullWidth type="text" label="Enter the Title or Author name"  onChange={this.onChange} id="outlined-size-small" variant="outlined"size="small"/>
                    {error === true && (
                      <div className="errormessage">title or author name is required</div>
                    )}
                  </Col>
                  <Col sm={6} md={4} lg={2} className="addIndent">
                    <Button type="submit" className="col-md-12" disabled={!this.state.value || this.state.value.trim().length === 0} style={{ backgroundColor: "violet",color:"white"}} variant="contained" block onClick={this.handleLogin}>Search</Button>
                  </Col>
                </Row>
              </Form>
            </CardBody>
          </Card>
          <div>
          </div>
        </Container>
        {BookContainer}
      
        {this.state.noData ? (
          <div><br/>
            <h5 className="centerIt">Sorry,we are not able to find the book you are looking for</h5>
            </div>
            
            ):(count>1 ? (
              <div>
                 <ReactPaginate
                    previousLabel={"prev"}
                    nextLabel={"next"}
                    breakLabel={"..."}
                    breakClassName={"break-me"}
                    pageCount={count}
                    marginPagesDisplayed={2}
                    pageRangeDisplayed={5}
                    onPageChange={this.handlePageClick}
                    containerClassName={"pagination"}
                    subContainerClassName={"pages pagination"}
                    activeClassName={"active"}/>
              </div>
            ):(<div></div>)
            )}
      </div>
    )
  }
}

export default Project;