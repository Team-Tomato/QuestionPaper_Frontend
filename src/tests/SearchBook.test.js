import React from 'react';
import {create} from 'react-test-renderer';
import { shallow, mount, render } from "enzyme";
import Loader from 'react-loading';
import { Form, Row, Col, Input, Button, Container, Card, CardBody, Table } from 'reactstrap'
import ReactPaginate from 'react-paginate';
import '../setupTest.js';
import '../Views/SearchBook.js';
import Project from '../Views/SearchBook.js';

describe('search book testing',()=>{
    test ("Matches the snapshot",()=>{
        const form = create(<Project />);
        expect(form.toJSON()).toMatchSnapshot(); 
    });

    test ("books list displayed when search button is clicked", ()=>{
        const handleLogin = jest.fn();
        const wrap = mount(<Button onClick = {handleLogin} />);
        const box = wrap.find('Button');
        box.simulate('click');
        expect(handleLogin.mock.calls.length).toEqual(1);
    });

    test ("title or author name is entered", ()=>{
        let input = {value: "data structure"};
        input['type'] = "text";
        expect(input).toEqual({value: "data structure", type: "text"});
    }); 
 
    test("should show error if any value is null", () => {
        expect(() => errormessage.toThrow())
    });

    
    test.skip('fetches specific book and displays', done => {
        const mockSuccessResponse = {};
        const mockJsonPromise = Promise.resolve(mockSuccessResponse);
        const mockFetchPromise = Promise.resolve({
            json: ()=> mockJsonPromise,
        });
        global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
        const wrapper = shallow(<Project />);
        expect(global.fetch).toBeCalled();
        expect(global.fetch).toHaveBeenCalledWith('https://teamtomato.herokuapp.com/api/v1/book/search?search_str=');
    });


});