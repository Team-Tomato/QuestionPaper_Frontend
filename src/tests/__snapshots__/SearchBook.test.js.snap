exports[`search book testing Matches the snapshot 1`] = `
<div>
  <div
    className="gradient card">
    <div
      className="welcome-title card-body"
      style={
        Object {
          "color": "white",
          "left": "50%",
          "position": "absolute",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        }
      }>
      <h5>
        Books destination for Integrated students
      </h5>
    </div>
  </div>
  <br />
  <div
    className="container">
    <div
      className="correctMargin card">
      <div
        className="removeIndent card-body">
        <form
          className="">
          <div
            className="row">
            <div
              className="addIndent col-sm-6 col-md-8 col-lg-10">
              <input
                className="form-control"
                onChange={[Function]}
                placeholder="Enter the Title or Author name"
                type="text" />
            </div>
            <div
              className="addIndent col-sm-6 col-md-4 col-lg-2">
              <button
                aria-label={null}
                className="btn btn-secondary btn-block disabled"
                disabled={true}
                onClick={[Function]}
                style={
                  Object {
                    "backgroundColor": "violet",
                  }
                }
                type="button"
                variant="contained">
                Search
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
    <div />
  </div>
  <h6 />
</div>
`;
